Duplicati/Library/Utility/ReadLimitLengthStream.cs

```diff
@@ -4,24 +4,66 @@
 using System.Threading.Tasks;
 
 namespace Duplicati.Library.Utility
-{
-    // StreamReadLimitLengthWrapper() based on code from Matt Smith (Oct 26 '15 at 20:38)
-    // https://stackoverflow.com/questions/33354822/how-to-set-length-in-stream-without-truncated
-
+{
+    // StreamReadLimitLengthWrapper() based on code from Matt Smith (Oct 26 '15 at 20:38)
+    // https://stackoverflow.com/questions/33354822/how-to-set-length-in-stream-without-truncated
+
+    /// <summary>
+    /// This class wraps a Stream but only allows reading a limited number of bytes from the underlying stream.
+    /// This can be used to create split a source stream into multiple smaller buffers that can be used as independent streams.
+    /// </summary>
     public class ReadLimitLengthStream : Stream
     {
-        readonly Stream m_innerStream;
-        readonly long m_endPosition;
-
-        public ReadLimitLengthStream(Stream innerStream, long size)
-        {
-            if (size < 0)
+        private readonly Stream m_innerStream;
+        private readonly long m_start;
+        private readonly long m_length;
+
+        public ReadLimitLengthStream(Stream innerStream, long length)
+            : this(innerStream, 0, length)
+        {
+        }
+
+        public ReadLimitLengthStream(Stream innerStream, long start, long length)
+        {
+            if (start < 0 || start > innerStream.Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(start));
+            }
+            if (length < 0 || innerStream.Length < start + length)
             {
-                throw new ArgumentOutOfRangeException(nameof(size));
+                throw new ArgumentOutOfRangeException(nameof(length));
             }
 
             m_innerStream = innerStream ?? throw new ArgumentNullException(nameof(innerStream));
```

Comment:
I think since we reference `innerStream.Length` above, `innerStream` can no longer be `null` here.

Reply:
Right - good catch. Fixed.